lib/pengine/pe_digest.c

```diff
@@ -45,30 +45,38 @@ pe__free_digests(gpointer ptr)
     }
 }
 
+/*!
+ * \internal
+ * \brief Remove named attributes from an XML element
+ *
+ * \param[in,out] param_set     XML to be filtered
+ * \param[in]     param_string  Space-separated list of attribute names
+ * \param[in]     need_present  Whether to remove attributes that match,
+ *                              or those that don't match
+ */
 static void
-filter_parameters(xmlNode * param_set, const char *param_string, bool need_present)
+filter_parameters(xmlNode *param_set, const char *param_string,
+                  bool need_present)
 {
-    if (param_set && param_string) {
-        xmlAttrPtr xIter = param_set->properties;
-
-        while (xIter) {
-            const char *prop_name = (const char *)xIter->name;
-            char *name = crm_strdup_printf(" %s ", prop_name);
-            char *match = strstr(param_string, name);
+    if ((param_set == NULL) || (param_string == NULL)) {
+        return;
+    }
+    for (xmlAttrPtr xIter = param_set->properties; xIter; ) {
+        const char *prop_name = (const char *) xIter->name;
+        char *name = crm_strdup_printf(" %s ", prop_name);
+        char *match = strstr(param_string, name);
 
-            free(name);
+        free(name);
 
-            //  Do now, because current entry might get removed below
-            xIter = xIter->next;
+        //  Do now, because current entry might get removed below
+        xIter = xIter->next;
 
-            if (need_present && match == NULL) {
-                crm_trace("%s not found in %s", prop_name, param_string);
-                xml_remove_prop(param_set, prop_name);
+        if ((need_present && (match == NULL))
+            || (!need_present && (match != NULL))) {
 
-            } else if (need_present == FALSE && match) {
-                crm_trace("%s found in %s", prop_name, param_string);
-                xml_remove_prop(param_set, prop_name);
-            }
+            crm_trace("Filtering %s (%sfound in '%s')",
+                      prop_name, (need_present? "not " : ""), param_string);
+            xml_remove_prop(param_set, prop_name);
```

Comment:
Having something like xml_remove_props as to not have to search twice might speed up things.
I know that it maps to xmlUnsetProp coming from libxml2.

Reply:
Good point, I think we should just use xmlRemoveProp() (which operates directly on an xmlAttrPtr) instead of xml_remove_prop(). Here we know we don't care about ACLs or diffs, so we can skip the wrappers.

Reply:
But even then we iterate twice over the list.

Reply:
No, xmlAttr is a doubly-linked list (and also has a link back to its parent element). xmlUnsetProp() would iterate a second time to find the attribute, but xmlRemoveProp() has everything needed

Reply:
sounds good - hadn't checked